src/groth16/circuit/builder/mod.rs

```diff
@@ -755,9 +913,333 @@ where
+    where
+        Z: IntoIterator<Item = &'a WireId> + BinaryInput + Copy,
+        P: IntoIterator<Item = &'a WireId> + BinaryInput + Copy,
+    {
+        let is_greater_than = self.greater_than(left, right);
+        let eq = self.is_equal(left, right);
+        self.new_or(is_greater_than, eq)
+    }
+
+    /// The WireId evaluates to one iff left > right
+    ///
+    /// ```
+    /// use zksnark::groth16::circuit::Circuit;
+    /// use zksnark::field::z251::Z251;
+    ///
+    /// let mut circuit = Circuit::<Z251>::new();
+    ///
+    /// let input_wire = circuit.new_word64();
+    /// let zero = circuit.const_word64(5);
+    /// let greater_than =
+    ///     circuit.greater_than(&input_wire, &zero);
+    ///
+    /// circuit.set_word64(&input_wire, 26);
+    /// assert_eq!(circuit.evaluate(greater_than), Z251::from(1));
+    ///
+    /// circuit.reset();
+    /// circuit.set_word64(&input_wire, 4);
+    /// assert_eq!(circuit.evaluate(greater_than), Z251::from(0));
+    ///
+    /// let input_wire = circuit.new_word8();
+    /// let zero = circuit.const_word64(5);
+    /// let greater_than =
+    ///     circuit.greater_than(&input_wire, &zero);
+    ///
+    /// circuit.set_word8(&input_wire, 26);
+    /// assert_eq!(circuit.evaluate(greater_than), Z251::from(1));
+    ///
+    /// circuit.reset();
+    /// circuit.set_word8(&input_wire, 4);
+    /// assert_eq!(circuit.evaluate(greater_than), Z251::from(0));
+    /// ```
+    pub fn greater_than<'a, Z, P>(&mut self, left: Z, right: P) -> WireId
+    where
+        Z: IntoIterator<Item = &'a WireId> + BinaryInput,
+        P: IntoIterator<Item = &'a WireId> + BinaryInput,
+    {
+        let mut l_iter = left.into_iter();
+        let mut r_iter = right.into_iter();
+        let left_first_wire = l_iter
+            .next()
+            .expect("greater_than: left input must have at least one WireId");
+        let right_first_wire = r_iter
+            .next()
+            .expect("greater_than: right input must have at least one WireId");
+
+        let mut cmp: VecDeque<WireId> = VecDeque::new();
+        let mut eq: VecDeque<WireId> = VecDeque::new();
+
+        // you or this at the end in case one of the extra left input
+        // is one.
+        // (think greater_than(Word64, Word8))
+        let mut left_extra_non_zero: WireId = self.const_wire_id(Binary::Zero);
+
+        // you new_and this at the end to make sure the right input's
+        // extra bits are all zero.
+        // (think greater_than(Word8, Word64))
+        let mut right_extra_all_zero: WireId = self.const_wire_id(Binary::One);
+        let cmp0 = self.new_greater_than(*left_first_wire, *right_first_wire);
+
+        l_iter.zip_longest(r_iter).for_each(|x| match x {
+            Both(l, r) => {
+                cmp.push_back(self.new_greater_than(*l, *r));
+                eq.push_back(self.new_equality(*l, *r));
+            }
+            Left(l) => {
```

Comment:
Again I think that this function should panic if the inputs are not the same length

Reply:
I agree, panicking is the best approach here.

Reply:
I think its much better for the comparison functions to work for all combinations of number types since the type system constrains the inputs for us to ensure all possible invocations of the function are meaningful. (See previous comment)

Reply:
Whatever we do to solve the other one can be done here.